lib/spack/spack/fetch_strategy.py

```diff
@@ -1425,8 +1601,14 @@ class FsCache(object):
     def __init__(self, root):
         self.root = os.path.abspath(root)
 
+    # TODO: use this method to determine the cache path for url fetches! This would
+    # invalidate all cached artifacts -- would our users allow this?
```

Comment:
I'm not entirely sure what this comment refers to.

Reply:
A few things:
1. `FsCache.persistent_cache_dir_for()` is now actually defined in a way that supports persistent caching of *any* fetcher (as per the prompt to reserve `svn` and `hg` subdirectories for future use).
1. Despite that, we don't ever use this method for `UrlFetchStrategy` instances, because they use the `.fetcher()` method immediately below which requires that the fetcher be a `UrlFetchStrategy`. Instead of having two methods to do the same thing (find a subdirectory of the cache root), it would be really nice if all fetchers could use the *same* method.
1. We *could* right now implement `.fetcher()` using the new `.persistent_cache_dir_for()` method, *but* if we did that, it would start caching `UrlFetchStrategy` fetches into `var/spack/cache/_source-cache/archive/url/ab/ab..../` instead of where it currently does, which is `var/spack/cache/_source-cache/archive/ab/ab.../` (note the `/url/` component in the first and not the second).
    - It would be much cleaner and easier to follow if they all used the same method `.persistent_cache_dir_for()`, but since that change would modify the cache paths, that would effectively invalidate the entire url fetch cache for spack users. We could do some clever symlinking to avoid this, but I didn't want to introduce that complexity and just left it as having the two separate duplicate methods in `.fetcher()` and `.persistent_cache_dir_for()`.

Reply:
I have reworded this to make it significantly more concise!